packages/office-ui-fabric-react/src/components/Button/_base/MenuButton.base.tsx

```diff
@@ -0,0 +1,117 @@
+import * as React from 'react';
+import {
+  BaseComponent,
+  autobind,
+  getId,
+  KeyCodes,
+  customizable
+} from '../../../Utilities';
+import { ButtonBase } from './Button.base';
+import { IButtonBaseProps } from './Button.base.types';
+import { DirectionalHint } from '../../../Callout';
+import { ContextualMenu, IContextualMenuProps } from '../../../ContextualMenu';
+import { IMenuButtonBaseProps } from './MenuButton.base.types';
+
+export interface IMenuButtonBaseState {
+  menuIsOpen: boolean;
+}
+@customizable('MenuButtonBase', ['theme'])
+export class MenuButtonBase extends BaseComponent<IMenuButtonBaseProps, IMenuButtonBaseState> {
+
+  private _labelId: string;
+
+  constructor(props: IMenuButtonBaseProps) {
+    super(props);
+    this._labelId = getId();
+
+    this.state = {
+      menuIsOpen: false
+    };
+  }
+
+  public render(): JSX.Element {
+    const {
+      getStyles
+    } = this.props;
+    return (
+      <ButtonBase
+        onClick={ this._onMenuClick }
+        onKeyDown={ this._onMenuKeyDown }
+        labelId={ this._labelId }
+        expanded={ this.state.menuIsOpen }
+        aria-expanded={ this.state.menuIsOpen }
+        {...this.props as IButtonBaseProps}
+        getStyles={ getStyles }
+        data-target-id={ this._labelId }
+        onRenderSuffix={ this._onRenderSuffix }
+      />
+    );
+  }
+
+  public dismissMenu(): void {
+    this.setState({ menuIsOpen: false });
+  }
+
+  @autobind
+  private _onRenderSuffix(): JSX.Element | null {
+    const {
+      menuProps,
+      onRenderMenu = this._onRenderMenu
+    } = this.props;
+
+    if (this.state.menuIsOpen && this.props.menuProps) {
+      return (
+        onRenderMenu(menuProps, this._onRenderMenu)
+      );
+    } else {
+      return null;
+    }
+  }
+
+  @autobind
+  private _onRenderMenu(menuProps: IContextualMenuProps): JSX.Element {
+    const {
+      onDismiss = this._onToggleMenu,
+      target = `[data-target-id='${this._labelId}']`
+    } = menuProps;
+
+    return (
+      <ContextualMenu
+        id={ this._labelId + '-menu' }
+        directionalHint={ DirectionalHint.bottomLeftEdge }
+        {...menuProps}
+        className={ 'ms-BaseButton-menuhost ' + menuProps.className }
+        target={ target }
+        labelElementId={ this._labelId }
+        onDismiss={ onDismiss }
```

Comment:
The onDismiss method should be moved from base button and used here instead of _onToggleMenu. I noticed a bug in Office Online where contextualmenu calls onDismiss twice when the document canvas takes focus (which we do when a menu is dismissed). This results in the menu opening and closing since onToggleMenu gets called twice. Also, onToggleMenu is not  a great name it should really be _toggleMenu, but that's not your fault :)

Reply:
I changed onToggle to toggle and removed the dissmissMenu from Button.base. Was there something else to do here?

Reply:
this should be done.